sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java

```diff
@@ -129,94 +130,144 @@ public void onError(String info, Exception e){
-							public void onComplete(boolean success) {
-								if (success) {
-									SetAppIcon msg = new SetAppIcon(SdlManager.this.appIcon.getName());
-									_internalInterface.sendRPCRequest(msg);
-								}
-							}
-						});
-					} else {
-						SetAppIcon msg = new SetAppIcon(SdlManager.this.appIcon.getName());
-						_internalInterface.sendRPCRequest(msg);
+	void checkState() {
+		if (permissionManager != null && fileManager != null && screenManager != null && (!lockScreenConfig.isEnabled() || lockScreenManager != null)) {
+			if (permissionManager.getState() == BaseSubManager.READY && fileManager.getState() == BaseSubManager.READY && screenManager.getState() == BaseSubManager.READY && (!lockScreenConfig.isEnabled() || lockScreenManager.getState() == BaseSubManager.READY)) {
+				DebugTool.logInfo("Starting sdl manager, all sub managers are in ready state");
+				transitionToState(BaseSubManager.READY);
+				notifyDevListener(null);
+				onReady();
+			} else if (permissionManager.getState() == BaseSubManager.ERROR && fileManager.getState() == BaseSubManager.ERROR && screenManager.getState() == BaseSubManager.ERROR && (!lockScreenConfig.isEnabled() || lockScreenManager.getState() == BaseSubManager.ERROR)) {
+				String info = "ERROR starting sdl manager, all sub managers are in error state";
+				Log.e(TAG, info);
+				transitionToState(BaseSubManager.ERROR);
+				notifyDevListener(info);
+			} else if (permissionManager.getState() == BaseSubManager.SETTING_UP || fileManager.getState() == BaseSubManager.SETTING_UP || screenManager.getState() == BaseSubManager.SETTING_UP || (lockScreenConfig.isEnabled() && lockScreenManager != null && lockScreenManager.getState() == BaseSubManager.SETTING_UP)) {
+				DebugTool.logInfo("SETTING UP sdl manager, some sub managers are still setting up");
+				transitionToState(BaseSubManager.SETTING_UP);
+				// No need to notify developer here!
+			} else {
+				Log.w(TAG, "LIMITED starting sdl manager, some sub managers are in error or limited state and the others finished setting up");
+				transitionToState(BaseSubManager.LIMITED);
+				notifyDevListener(null);
+				onReady();
+			}
+		} else {
+			// We should never be here, but somehow one of the sub-sub managers is null
+			String info = "ERROR one of the sdl sub managers is null";
+			Log.e(TAG, info);
+			transitionToState(BaseSubManager.ERROR);
+			notifyDevListener(info);
+		}
+	}
+
+	private void notifyDevListener(String info) {
+		if (managerListener != null) {
+			if (getState() == BaseSubManager.ERROR){
+				managerListener.onError(info, null);
+			} else {
+				managerListener.onStart();
+			}
+		}
+	}
+
+	private void onReady(){
+		// Set the app icon
+		if (SdlManager.this.appIcon != null && SdlManager.this.appIcon.getName() != null) {
+			if (fileManager != null && (fileManager.getState() == BaseSubManager.READY || fileManager.getState() == BaseSubManager.LIMITED) && !fileManager.hasUploadedFile(SdlManager.this.appIcon)) {
```

Comment:
i don't think file manager is ever in a limited state

Reply:
Makes sense. I will remove that check.